### DISCUSSION TITLE
Roth oil tank analog gauge

### DISCUSSION THREAD

**Discussion by benb13** [jomjol/AI-on-the-edge-device#3020] [General] [answered]
I am trying to use the analog reading on a Roth oil tank gauge. I am wondering if anyone has adjusted the water meter model or trained a new model for this type of analog gauge. Thank you.
![alg_roi](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/348f09df-0709-4f6d-9ef9-3a37af5ffb74)

**Comment by caco3** [comment]
The analog model tries to find the angle of the pointer. Then it scales the given angle from 360° to 1.0.
So if you success to get a proper reading and accept the need to do post processing, you could convert the value back to an angle resp. the effective value.

Looking on your screenshot you need to improve the focus. also the alignment marks need to be as far away as possible. Use stickers eg. on the right bottom and top left corner.

**Reply by benb13** [reply]
Thank you for your recommendations. I will give that a try and update the post.

**Reply by benb13** [reply]
I was able to make both changes you suggested. I added a sticker to the top left and bottom right and updated the markers. I also found an article on how to adjust the focus, so I have a much clearer image. Now I am working on seeing if we can get valid readings.
![alg_roi2](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/3e054121-5636-4848-9201-ab1e069bf07e)
![download2](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/4f6729fe-db06-48ec-b2ab-f3fb884275b6)

**Reply by caco3** [reply]
That looks good now.

Just make sure to set [AllowNegativeRates](https://jomjol.github.io/AI-on-the-edge-device-docs/Parameters/#parameter-numbersallownegativerates) to true!

**Reply by benb13** [reply]
Thank you. I read through all of the documentation and that was the first thing I changed since the gauge will be decrementing.  It now does not give me a negative rate error like above. It has been running for about a day (every 30 minutes) with the new focus and settings. I don't believe I am getting a correct reading. The RAW readings are between 4 and 6. It has nothing in the value setting. i am assuming the post processing is failing. What am I supposed to see if the reading is correct? It is supposed to be between 0 and 1 or the angle of the gauge? I am not sure what the 4, 5 or 6 are for the raw value. I appreciate all of your help so far.
![Screenshot 2024-04-04 090248](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/fd3b9a8d-c521-4b2f-b453-75ada6a0ef64)
![Screenshot 2024-04-04 090320](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/ebbd776a-dba1-4356-ac69-a6acb40f7d6d)

**Reply by caco3** [reply]
As you can see, you still get an error: "rate to high".
You need to:
1. set the max rate to a large value or disable it for now.
1. if there is still an error, once set the `preValue` to the current value (approx. 0.2 on the screenshot).

A pointer interprets the full rotation of 0..360° as 0..1, rotating clock wise (unless explicitely configured).
since you only use the top halve, you only should get values from 0..2.5 and 7.5..9.9
so you will have to convert the calculated values back to your 0..1 scale:
|AIOTE Interpretation|Meter Scale|
|--|--|
|7.5|0|
|8.75|0.25|
|0.0|0.5|
|1.25|0.75|
|2.5|1.0|

**Reply by benb13** [reply]
Thank you! That helps a lot to understand the range I should be looking for with the values. I made the change to the max rate, which removed the error, but the readings are still within the same range of 5 to 6. Those values are in the lower to lower left quadrant so it does not seem correct.

Here is a sampling:
2024-04-05T03:24:29-0400,main,4,4,4,-0.033352,-1,no error,4.9
2024-04-05T03:54:29-0400,main,5,5,5,0.033333,1,no error,5.2
2024-04-05T04:24:30-0400,main,5,5,5,0.000000,0,no error,5.2
2024-04-05T04:54:30-0400,main,5,5,5,0.000000,0,no error,5.0
2024-04-05T05:24:30-0400,main,5,5,5,0.000000,0,no error,5.3
2024-04-05T05:54:30-0400,main,4,4,4,-0.033333,-1,no error,4.7
2024-04-05T06:24:29-0400,main,5,5,5,0.033352,1,no error,5.9
2024-04-05T06:54:29-0400,main,5,5,5,0.000000,0,no error,5.5
2024-04-05T07:24:29-0400,main,5,5,5,0.000000,0,no error,5.5
2024-04-05T07:54:29-0400,main,6,6,6,0.033333,1,no error,6.2
2024-04-05T08:24:29-0400,main,5,5,5,-0.033333,-1,no error,5.6
2024-04-05T08:54:29-0400,main,5,5,5,0.000000,0,no error,5.9
2024-04-05T09:24:29-0400,main,5,5,5,0.000000,0,no error,5.4
2024-04-05T09:54:29-0400,main,5,5,5,0.000000,0,no error,5.2
2024-04-05T10:24:29-0400,main,4,4,4,-0.033333,-1,no error,4.9

I am thinking that the oil gauge may not be similar enough to the water meter and giving inaccurate readings. I had some thoughts to maybe tighten the ROI so that it includes less white space and more of the needle. Other thoughts are to put a sticker over the red ROTH letters since they are possibly being seen as the red needle for the water meter model. I am going to try some ideas.

I am open to any other suggestions. This seems like amazing software.

**Reply by benb13** [reply]
Interesting. tightening the ROI is giving me new readings.

2024-04-05T10:49:55-0400,main,3,3,3,-0.039318,-1,no error,3.1
2024-04-05T11:19:55-0400,main,2,2,2,-0.033333,-1,no error,2.6
2024-04-05T11:25:37-0400,main,3,3,3,0.175439,1,no error,3.1

![Screenshot 2024-04-05 112735](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/0fb3ba69-2bb0-4af3-9f0f-968e456d49a0)

**Reply by caco3** [reply]
you could make it even smaller, so that no red parts are contained.
but I also think that you need to train your own model!

**Reply by benb13** [reply]
Thanks. I was thinking the same that I might need to train the model with pics. I will start looking at that documentation. I did take pictures with my phone the last time they filled the tank. I wonder if I can use them to train a model.
![image](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/c47c4959-77eb-436c-b85b-954d31a4c005)

**Comment by benb13** [comment]
Removed my last post about the notebook run errors. I found documentation in the repository for a windows install. I will give that a try and post back.

**Reply by SybexX** [reply]
It's best to use Anaconda https://www.anaconda.com/download

**Comment by benb13** [comment]
Thank you. I am going through that installation and setup now using steps in Training-Setup-Windows.md.

**Reply by SybexX** [reply]
With Anaconda it is relatively easy and there is a lot of information and videos about installing and setting up.
e.g. this video https://www.youtube.com/watch?v=WUeBzT43JyY

**Reply by benb13** [reply]
@SybexX Thank you for the links. It was an easy setup to load Anaconda and all the dependencies. Next I was able to successfully run the **02 - Image_Preparation.ipynb** notebook and it created all of the resized images in the data_resize_all folder. I then loaded the **03 - Train_CNN_Analog-Readout_Version-Small2.ipynb** notebook and ran it. It seemed to complete, but I have no output of an updated model. I even tried to change the model version and I don't see an updated model anywhere in the directories. I looked through the output messages and it does not look like there were any errors. I am assuming it writes the new model to the root directory with the filename format specified in the code. I am attaching some screen shots of the code that was run. It looks like I am close. Just trying to figure out where to find the updated model or where to find any errors with the model creation.

![Screenshot 2024-04-23 073456](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/1293b74f-7fc3-4062-b535-4b3fed1c546f)
![Screenshot 2024-04-23 073923](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/921321f7-32ce-4939-90d3-09ad88e8555c)
![Screenshot 2024-04-23 073955](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/a2774cfc-c1e3-4a18-a6ec-ff1c6667648a)
![Screenshot 2024-04-23 074027](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/0fb8900d-6830-431a-a515-4931a98bf0f4)
![Screenshot 2024-04-23 074109](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/c25e39d8-67c3-44ff-9b0c-a303a3472406)

**Reply by SybexX** [reply]
Since I don't see any output windows in the last image, you obviously didn't follow these steps either.

**Reply by benb13** [reply]
@SybexX Thank you. I am new to Anaconda and it was not obvious to me when each step was completed. This was my fault. I did not wait until the 1000 iteration process ended before I clicked to run the next step. I then stepped more carefully through the notebook and waited for the kernel to finish and.... I have a compiled model! 

I copied it over to my esp32 last night and it is running. The first couple values looked good. They match the sample images and within the range I was expecting. I updated my configuration to allow for precision values. I am going to run it for a while and monitor the values. My last step I will work on is translating the values of 7.5-9.9/0-2.5 into 0-100.

@SybexX @caco3 I really appreciate all of your help. I have some warnings and errors to work through, but I couldn't have got this far without your direction. This is amazing software.

![Screenshot 2024-04-23-01](https://github.com/jomjol/AI-on-the-edge-device/assets/5447023/f4e041db-c197-430a-9e5e-a6579cb8d7d4)

**Reply by caco3** [reply]
For a further test with your model, you could use the demo modus of the device with your own images, see https://jomjol.github.io/AI-on-the-edge-device-docs/Demo-Mode/

**Comment by benb13** [comment, marked as answer]
To use AI on the Edge for different analog gauges, the recommendation can be to train your own model using the tools and documentation provided.
1) Collect sample images for your different gauge values, or start the AI on the Edge config to save images for creating your own model.
2) Clear the directory and copy your images to data_raw_all
3) Following guides under Training-Setup-Windows.md, Training_Data.md and/or Train_Network.md
4) Run the Notebooks in the order they are numbered.
5) Once created successfully, copy the .flite files to your esp32 SD and update your config to reflect the custom model.
